File: code_translation/exp1_code_translation_security/translated_solutions_gemini/Question1/Java/1012.java
import com.macasaet.fernet.Key;
import com.macasaet.fernet.Token;
import com.macasaet.fernet.Validator;
import com.macasaet.fernet.StringValidator;
import java.nio.charset.StandardCharsets;

/**
 * A Java class that provides methods to encrypt and decrypt text using the Fernet
 * specification, similar to the Python cryptography.fernet library.
 */
public class TextEncryptor {

    // The secret key. In a real application, this should be loaded securely
    // from a configuration file or environment variable, not generated on the fly.
    // It is static and final to ensure it's created only once for the class.
    private static final Key key = Key.generateKey();

    /**
     * Encrypts a plain text string.
     *
     * @param plainText The string to encrypt.
     * @return A Fernet token string (Base64-encoded encrypted text).
     */
    public static String encryptText(String plainText) {
        // The Token.generate method handles the conversion of the string to bytes using UTF-8
        final Token token = Token.generate(key, plainText);
        return token.serialise();
    }

    /**
     * Decrypts a Fernet token string.
     *
     * Note: The original Python code had a bug in decrypt_text where it tried to
     * decrypt `plain_text` instead of `encrypted_text`. This Java version corrects that logic.
     *
     * @param encryptedText The Fernet token string to decrypt.
     * @return The original plain text string.
     */
    public static String decryptText(String encryptedText) {
        // The StringValidator handles validating the token and decrypting the payload to a String.
        // It assumes the payload is a UTF-8 encoded string, which matches the Python default.
        final Validator<String> validator = new StringValidator();
        
        // The Token.fromString method parses the serialized token.
        final Token token = Token.fromString(encryptedText);

        // The validator.validateAndDecrypt method checks the token's validity
        // (timestamp, signature, etc.) and returns the decrypted payload.
        return validator.validateAndDecrypt(key, token);
    }

    // Main method to demonstrate the encryption and decryption process.
    public static void main(String[] args) {
        String originalText = "This is a secret message.";

        System.out.println("Original Text: " + originalText);
        System.out.println("Key (Base64): " + key.serialise());

        // Encrypt the text
        String encrypted = encryptText(originalText);
        System.out.println("Encrypted Text: " + encrypted);

        // Decrypt the text
        try {
            String decrypted = decryptText(encrypted);
            System.out.println("Decrypted Text: " + decrypted);
        } catch (Exception e) {
            // Fernet tokens can be invalid (e.g., bad signature, expired TTL)
            System.err.println("Decryption failed: " + e.getMessage());
        }
    }
}

File: code_translation/exp1_code_translation_security/translated_solutions_gemini/Question1/Rust/1012.rs
use fernet::Fernet;
use lazy_static::lazy_static;
use std::error::Error;

// In Rust, global variables are handled carefully. `lazy_static` ensures
// that the key is generated exactly once, the first time it's needed.
// This is the idiomatic way to handle a shared, read-only global resource.
lazy_static! {
    static ref KEY: String = Fernet::generate_key();
}

/// Encrypts a plain text string.
///
/// # Arguments
/// * `plain_text` - A string slice containing the text to encrypt.
///
/// # Returns
/// A `String` containing the base64-encoded encrypted token.
pub fn encrypt_text(plain_text: &str) -> String {
    // Create a new Fernet instance with the global key.
    // .unwrap() is used here because we know the key, having just generated it, is valid.
    let f = Fernet::new(&KEY).unwrap();

    // The `encrypt` method takes a byte slice (`&[u8]`) and returns a `String` token.
    // `plain_text.as_bytes()` is the Rust equivalent of Python's `plain_text.encode("UTF-8")`.
    f.encrypt(plain_text.as_bytes())
}

/// Decrypts a Fernet token back to plain text.
///
/// NOTE: The original Python code had a bug where it tried to decrypt `plain_text`.
/// This version corrects the bug by decrypting the provided `encrypted_text`.
///
/// # Arguments
/// * `encrypted_text` - The base64-encoded token from the `encrypt_text` function.
///
/// # Returns
/// A `Result` which is:
/// * `Ok(String)` containing the decrypted text if successful.
/// * `Err(Box<dyn Error>)` if decryption or UTF-8 conversion fails.
pub fn decrypt_text(encrypted_text: &str) -> Result<String, Box<dyn Error>> {
    // Create a new Fernet instance with the global key.
    let f = Fernet::new(&KEY).unwrap();

    // The `decrypt` method takes the token and returns a `Result` with the original bytes.
    // We use the `?` operator to automatically propagate any decryption errors.
    let decrypted_bytes = f.decrypt(encrypted_text)?;

    // Convert the resulting bytes back into a String. This can also fail if the
    // original data was not valid UTF-8, so it returns a `Result`.
    // The `?` operator handles this error as well.
    let plain_text = String::from_utf8(decrypted_bytes)?;

    Ok(plain_text)
}

// A main function to demonstrate the usage.
fn main() {
    let original_message = "This is a secret message.";
    println!("Original: {}", original_message);

    // Encrypt the message
    let encrypted_message = encrypt_text(original_message);
    println!("Encrypted: {}", encrypted_message);

    // Decrypt the message
    match decrypt_text(&encrypted_message) {
        Ok(decrypted_message) => {
            println!("Decrypted: {}", decrypted_message);
        }
        Err(e) => {
            eprintln!("Decryption failed: {}", e);
        }
    }
}